File: backend/domain/schemas/forecast.py
"""
Pydantic schemas for citation forecast API responses.
"""

from pydantic import BaseModel
from typing import Literal, Optional


# ── Shared ──────────────────────────────────────────────────────────────────

class Interval80(BaseModel):
    low: float
    high: float


class ForecastMeta(BaseModel):
    model_version: str
    calibration: str = "bucketed_conformal_80"
    as_of_definition: str = "filing_plus_2y"


# ── Patent-level ────────────────────────────────────────────────────────────

class PatentPrediction(BaseModel):
    expected_citations: float
    interval_80: Interval80
    difficulty_bucket: Literal["LOW", "MID", "HIGH"]


class PatentForecastResponse(BaseModel):
    appln_id: int
    horizon: str
    filing_date: str
    as_of_date: str
    prediction: PatentPrediction
    features_used: dict
    meta: ForecastMeta


# ── Portfolio-level ─────────────────────────────────────────────────────────

class DifficultyMix(BaseModel):
    LOW: float
    MID: float
    HIGH: float


class TopContributor(BaseModel):
    appln_id: int
    owner_share: float
    expected: float
    interval_80: Interval80
    contribution: float
    difficulty_bucket: str


class PortfolioPrediction(BaseModel):
    expected_citations_total: float
    interval_80_total: Interval80
    n_patents_effective: float
    expected_per_effective_patent: float


class SegmentItem(BaseModel):
    key: str
    expected: float
    interval_80: Interval80
    share: float


class SegmentsMeta(BaseModel):
    top_k: int
    cpc_top_n_per_patent: int
    allocation: str
    other_bucket_label: str = "OTHER"


class PortfolioForecastSegments(BaseModel):
    meta: SegmentsMeta
    by_cpc_subclass: list[SegmentItem] = []
    by_jurisdiction_coverage_bucket: list[SegmentItem] = []
    by_major_office_grant_bucket: list[SegmentItem] = []
    by_as_of_year_bucket: list[SegmentItem] = []


class PortfolioForecastResponse(BaseModel):
    owner_id: int
    horizon: str
    as_of_date: str
    portfolio_prediction: PortfolioPrediction
    difficulty_mix: DifficultyMix
    top_contributors: list[TopContributor]
    segments: Optional[PortfolioForecastSegments] = None
    meta: ForecastMeta


class CitationForecastTimeSeriesItem(BaseModel):
    year: int
    cum_cites: float
    low: Optional[float] = None
    high: Optional[float] = None


class ForecastPredictionSummary(BaseModel):
    expected_additional: float
    interval_80: Interval80
    difficulty_bucket: str


class CitationForecastTimeSeriesResponse(BaseModel):
    appln_id: int
    filing_year: int
    horizon: str
    last_observed_year: int
    historical: list[CitationForecastTimeSeriesItem]
    forecast: list[CitationForecastTimeSeriesItem]
    prediction_summary: ForecastPredictionSummary

File: frontend_v1/lib/types/forecast.ts
/**
 * Citation Forecast API Types
 */

export type Interval80 = {
    low: number;
    high: number;
};

export type DifficultyBucket = "LOW" | "MID" | "HIGH";

// ── Patent Forecast ──────────────────────────────────────────────────────

export type PatentForecastPrediction = {
    expected_citations: number;
    interval_80: Interval80;
    difficulty_bucket: DifficultyBucket;
};

export type ForecastMeta = {
    model_version: string;
    calibration: string;
    as_of_definition: string;
    aggregation?: string;
};

export type PatentForecastResponse = {
    appln_id: number;
    horizon: string;
    filing_date: string;
    as_of_date: string;
    prediction: PatentForecastPrediction;
    features_used: Record<string, number | boolean>;
    meta: ForecastMeta;
};

// ── Portfolio Forecast ───────────────────────────────────────────────────

export type PortfolioForecastPrediction = {
    expected_citations_total: number;
    interval_80_total: Interval80;
    n_patents_effective: number;
    expected_per_effective_patent: number;
};

export type DifficultyMix = {
    LOW: number;
    MID: number;
    HIGH: number;
};

export type TopContributor = {
    appln_id: number;
    owner_share: number;
    expected: number;
    interval_80: Interval80;
    contribution: number;
    difficulty_bucket: DifficultyBucket;
};

export type SegmentItem = {
    key: string;
    expected: number;
    interval_80: Interval80;
    share: number;
};

export type SegmentsMeta = {
    top_k: number;
    cpc_top_n_per_patent: number;
    allocation: string;
    other_bucket_label: string;
};

export type PortfolioForecastSegments = {
    meta: SegmentsMeta;
    by_cpc_subclass: SegmentItem[];
    by_jurisdiction_coverage_bucket: SegmentItem[];
    by_major_office_grant_bucket: SegmentItem[];
    by_as_of_year_bucket: SegmentItem[];
};

export type PortfolioForecastResponse = {
    owner_id: number;
    horizon: string;
    as_of_date: string;
    portfolio_prediction: PortfolioForecastPrediction;
    difficulty_mix: DifficultyMix;
    top_contributors: TopContributor[];
    segments?: PortfolioForecastSegments | null;
    meta: ForecastMeta;
};

export type CitationForecastTimeSeriesPoint = {
    year: number;
    cum_cites: number;
    low?: number;
    high?: number;
};

export type CitationForecastTimeSeriesResponse = {
    appln_id?: number;
    owner_id?: number;
    filing_year: number;
    horizon: string;
    last_observed_year: number;
    historical: CitationForecastTimeSeriesPoint[];
    forecast: CitationForecastTimeSeriesPoint[];
    prediction_summary: {
        expected_additional: number;
        interval_80: Interval80;
        difficulty_bucket: DifficultyBucket | string;
    };
};
